### PR TITLE
Create instructions for loading HSL and Pardiso libraries at runtime

### PR DESCRIPTION
**Pull request by LukasBarner**
x-ref: https://github.com/jump-dev/Ipopt.jl/issues/342
Maybe it might make sense to run a conditional and something like this from inside the wrapper?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -182,6 +182,24 @@ To improve performance, Ipopt supports a number of linear solvers. Installing
 these can be tricky, however, the following instructions should work. If they
 don't, or are not explicit enough, please open an issue.
 
+### General Remarks
```
Comment (odow) [inline README.md:185]:
```suggestion
### General Remarks

```

**Review by odow — COMMENTED** [review]
Is this not just a problem with how we're configuring and compiling the linear solvers?

I don't think we want to do this inside Ipopt.jl, because it might have unintended consequences.

cc @amontoison

**Comment by amontoison** [comment]
> Is this not just a problem with how we're configuring and compiling the linear solvers?
> 
> I don't think we want to do this inside Ipopt.jl, because it might have unintended consequences.
> 
> cc @amontoison

It's an issue with how the shared libraries of the linear solvers are compiled.
Something is not correctly linked when the shared library `libcoinhsl` is generated.
You can regenerate the shared library from the static library with this command:
```shell
# Linux
gfortran -fPIC -shared -Wl,--whole-archive libcoinhsl.a -Wl,--no-whole-archive
                                        -lblas -llapack -lmetis -lgomp
                                        -o libcoinhsl.so

# MacOS
gfortran -fPIC -shared -Wl,-all_load libcoinhsl.a -Wl,-noall_load
                                        -lblas -llapack -lmetis -lgomp
                                        -o libcoinhsl.dylib
```

**Comment by LukasBarner** [comment]
Thanks for the quick feedback @amontoison 😄. I tried this and got the following error: 
```
gfortran -fPIC -shared -Wl,-all_load libcoinhsl.a -Wl,-noall_load -lblas -llapack -lmetis -lgomp -o libcoinhsl.dylib

ld: warning: option -noall_load is obsolete and being ignored
ld: library not found for -lmetis
collect2: error: ld returned 1 exit status
```
Which confuses me, because something like this works: 
```julia
julia> using Libdl

julia> dlopen("libmetis.dylib",RTLD_GLOBAL)
Ptr{Nothing} @0x00007fadce505080
```
Do you have any idea why metis is not found?

**Comment by amontoison** [comment]
> Thanks for the quick feedback @amontoison smile. I tried this and got the following error:
> 
> ```
> gfortran -fPIC -shared -Wl,-all_load libcoinhsl.a -Wl,-noall_load -lblas -llapack -lmetis -lgomp -o libcoinhsl.dylib
> 
> ld: warning: option -noall_load is obsolete and being ignored
> ld: library not found for -lmetis
> collect2: error: ld returned 1 exit status
> ```
> 
> Which confuses me, because something like this works:
> 
> ```julia
> julia> using Libdl
> 
> julia> dlopen("libmetis.dylib",RTLD_GLOBAL)
> Ptr{Nothing} @0x00007fadce505080
> ```
> 
> Do you have any idea why metis is not found?

No, is `libmetis.dylib` in a folder of `DYLD_FALLBACK_LIBRARY_PATH`?
You can help the linker otherwise:
```
gfortran -fPIC -shared -Wl,-all_load libcoinhsl.a -lblas
               -llapack -L/path/to/metis -lmetis -lgomp -o libcoinhsl.dylib
```

**Comment by odow** [comment]
Is a better solution to tell people to use https://github.com/JuliaSmoothOptimizers/HSL.jl?

**Comment by LukasBarner** [comment]
After building from http://glaros.dtc.umn.edu/gkhome/fsroot/sw/metis/OLD, [this](https://github.com/jump-dev/Ipopt.jl/pull/345#issuecomment-1331024042) also works without any additional specifications/tweaks! (except specifying `set_optimizer_attribute()`)
Is there an easier way to get metis 4 for osx than the above (homebrew only seems to work for 5)?

**Comment by amontoison** [comment]
I think that the best solution is that compile `libcoinhsl` with HSL.jl such that all dependencies are Julia artifacts.
I recompiled `METIS4` with a different soname in Yggdrasil and I directly link with it in HSL.jl.
https://github.com/JuliaSmoothOptimizers/HSL.jl/pull/150

**Comment by LukasBarner** [comment]
That's nice! The dlopen() way is really not a good idea... 
I believe that https://github.com/jump-dev/Ipopt.jl/issues/341 was caused by this.
It might make sense to also use the option of ipopt to load the pardiso library at runtime (I could try testing / creating instructions for this as well)...

**Comment by amontoison** [comment]
I think that your segmentation fault is another problem but all solvers will be updated with the new `coinhsl` archive so it could solve the issue.
The shared library will be libhsl.so also, so we will not need to rename it anymore.

We will also have the `c` and `z` versions of the solvers. I asked it for HSL.jl.

**Comment by odow** [comment]
Closing because we're about to revamp how this all works.

**Pull request closed without merging**